src/functions/data.js

```diff
@@ -0,0 +1,62 @@
+import defineFunction, {ordargument} from "../defineFunction";
+import buildCommon from "../buildCommon";
+import mathMLTree from "../mathMLTree";
+import ParseError from "../ParseError";
+
+import {assertNodeType} from "../parseNode";
+
+import * as html from "../buildHTML";
+import * as mml from "../buildMathML";
+
+defineFunction({
+    type: "data",
+    names: ["\\data"],
+    props: {
+        numArgs: 2,
+        allowedInText: true,
+        argTypes: ["text", "original"],
+    },
+    handler: ({parser}, args) => {
+        const body = args[1];
+
+        const attribGroup = args[0];
+        if (attribGroup.type !== "ordgroup") {
+            throw new ParseError("Invalid data attributes", attribGroup);
+        }
+        let dataAttributes = "";
+        for (let i = 0; i < attribGroup.body.length; ++i) {
+            if (attribGroup.body[i].type === "spacing") {
+                continue;
+            }
+            dataAttributes += assertNodeType(attribGroup.body[i], "textord").text;
+        }
+        dataAttributes = dataAttributes.split(',');
+        dataAttributes = dataAttributes.map(p => p.split('='));
+        dataAttributes = dataAttributes.reduce((obj, val) => {
+            obj[val[0]] = val[1] || '';
+            return obj;
+        }, {});
+
+        return {
+            type: "data",
+            mode: parser.mode,
+            body: ordargument(body),
+            dataAttributes,
+        };
+    },
+    htmlBuilder: (group, options) => {
+        const elements = html.buildExpression(
+            group.body,
+            options.withDataAttributes(group.dataAttributes),
```

Comment:
I can try doing that inline here. That does make it nicer to not pipe things through and would render the other comment about using `attributes` moot (though that is a good suggestion).

Regarding descendent elements, we actually did want that in our case. For example if we render `\data{a=b}{\frac{3}{4}}` we want the `3`, the `4`, and the fraction line to all have `data-a="b"` on them. We also wanted to support things like `\data{a=b}{\frac{\data{a=c, d=e}{3}}{4}}` where children can override and extend the set of data attributes.

Reply:
I can definitely take a look at this and see what makes sense. In my basic testing `buildExpression` was returning `SymbolNode`s which don’t have attribute support by default so that would still need to be added. I also want to make sure things like fraction bars and other visible elements get the appropriate attributes. But assuming I can get that working it does make sense to have this layer handle all of the data attributes so it’s less tangled into the general code base.

Reply:
I've tried this out and it's close, but I end up with two options, neither works for our scenarios:

1. We iterate through `elements` and their children, applying the attributes. This gets our attributes applied to all children but also applies them to non-visual DOM elements used for spacing and alignment which is not good for us since it makes those invisible elements considered part of the interactive DOM.
2. We iterate through `elements` without looking at children and apply the attributes. This solves the above problem but then naturally means for things like `\data{modelid=1}{\frac{3}{4}}` we don't get our attribute applied to the 3, 4, or the fraction bar which is something we want so that we can tie all three of those visual elements back to our single model ID.

Without writing a series of special cases in the data function handler (e.g. apply the attribute to elements that have `.text` or one of a set of special class names), I'm unsure of a way to have it implement this logic without using Options. I can make the Options object have a more generic `attributes` instead of `dataAttributes` and wire it through the existing attribute handling code, but that's the only way I've found to get the full behavior we need for interactivity.